### ISSUE TITLE
Wrong PHP version, OSM Wiki is down - downgrade below 7.4 and above 7.3.19 needed or upgrade above 7.4.8

### ISSUE THREAD

**Issue by matkoniecz** [openstreetmap/operations#511]
> MediaWiki 1.35 requires PHP 7.3.19 or higher (and not 7.4.0 - 7.4.8); you are using PHP 7.4.3.
> Supported PHP versions

> Please consider upgrading your copy of PHP. PHP versions less than v7.3.0 are no longer supported by the PHP Group and will not receive security or bugfix updates.

> If for some reason you are unable to upgrade your PHP version, you will need to download an older version of MediaWiki from our website. See our compatibility page for details of which versions are compatible with prior versions of PHP. 

https://wiki.openstreetmap.org/wiki/Key:amenity

**Comment by matkoniecz** [comment]
I checked https://github.com/openstreetmap/operations/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc to see is it case of PHP upgrade in progress that entered forbidden zone and will upgrade into supported version.

Not sure where I should also look to check for that

**Comment by tomhughes** [comment]
This is outside our control - we are using the Ubuntu supplied version of PHP and when we upgraded mediawiki to 1.35 it complied with all the requirements.

**Comment by matkoniecz** [comment]
Maybe it would be worth considering fact that OSM Wiki is down (even for reading) as a bug?

**Comment by tomhughes** [comment]
THat may well be the case but there is literally nothing I can reasonably do to fix it.

It sounds like this is basically enemy action by Mediawiki - they have, in a minor patch to 1.35, bumped the required PHP version with a hard failure if it's not available.

I have no way to upgrade PHP without extensive work and no reaosnable way to override the MW error.

**Comment by matkoniecz** [comment]
Is it possible to downgrade Mediawiki as a workaround around surprise breaking change? Or is it also unreasonably complicated?

> they have, in a minor patch to 1.35, bumped the required PHP version with a hard failure if it's not available.

WTF

**Comment by tomhughes** [comment]
T270228 is the phabricator reference I think.

Our chef recipes automatically track the upstream head and I have no easy way to override that.

**Comment by tomhughes** [comment]
I've patched out the blacklisting for now but that will only hold as long as chef is stopped.

**Comment by matkoniecz** [comment]
https://phabricator.wikimedia.org/T270228 - this one, right?

**Comment by tomhughes** [comment]
Yes.

**Comment by tomhughes** [comment]
Specifically https://gerrit.wikimedia.org/r/c/mediawiki/core/+/666700/ is them cherry picking it to 1.35

**Comment by tomhughes** [comment]
If I'm reading the ticket right downgrading MW doesn't help - the bug was always there but rather than fix it they just banned the PHP versions that trigger it.

**Comment by jdforrester** [comment]
Sorry about this; we just noticed a few minutes ago that Ubuntu haven't upgraded their PHP version in the 20.04 LTS (but have a compliant one in 20.10).

**Comment by matkoniecz** [comment]
> https://phabricator.wikimedia.org/T270228 - this one, right?

They replied that:

> There's been no "minor release" yet. You're following a development/release branch.

maybe switching to using released versions would be a good idea? Or is it workaround for another problem?

**Comment by tomhughes** [comment]
Well when we wrote that we understood that the branch only got updates as things were released.

If there's a better way to easily track things so that we only get actual release that would be good.

What I don't want is anything that relies on us manually bumping versions - with PHP stuff we really want to be pulling security fixes as soon as they're available without human interaction.

**Comment by tomhughes** [comment]
As far as I can see there aren't any release tags in the repository - we track the `REL1_35` branch which as I say was done because we thought that only got updated when a release was made.

I did, relatively recently, become aware that it probably did include edits that would be in the next release but short of retooling everything to work by downloading zips (which is quite a headache with PHP apps that mix user data with code) there wasn't any clear way to resolve that.

Even if I somehow reworked it to use the zips I would need some programratic way to identify the latest release for a given major version - wordpress actually have an API for that which we use, and tags in their repo which we can sync to, but as far as I know there is no equivalent for mediawiki.

**Comment by tomhughes** [comment]
Actually the gerrit repos (which is what we use in chef) do have tags, it's just the phabricator repos which do.

I guess I'll need to hack up something to parse `git ls-remote` and figure out the most recent tag and then sync to that.

**Comment by reedy** [comment]
>Actually the gerrit repos (which is what we use in chef) do have tags, it's just the phabricator repos which do.

Yes they do; they are what are used to generate the tarballs. Gerrit is the canonical repo.

You can see them on https://gerrit.wikimedia.org/g/mediawiki/core

https://gerrit.wikimedia.org/r/plugins/gitiles/mediawiki/core/+/refs/tags/1.35.1

And then these are the same as then appear over at Github too - https://github.com/wikimedia/mediawiki/tags

**Comment by matkoniecz** [comment]
Not sure is it worth copying that here (feel free to delete/edit/hide/downvote that comment if not) but https://phabricator.wikimedia.org/T270228#6863197 has following comment:

> However, you do bring up a couple of good points; and one that @Jdforrester-WMF mentioned to me on IRC... Maybe we should have a tag/branch pointer that points to the "latest released" version. So it points at 1.35.1 now, and when we release 1.35.2, it is moved to point there instead.

> And/or the other option is having some sort of machine readable API (I don't think we have anything currently) where you can poll it and get the latest point release for a branch. Then be able to use that in your chef recipe to be git checkout'd

> I'm guessing you're not expecting your wiki to automatically upgrade to 1.36 when that is released; ie that'd be a manual action?

**Comment by Firefishy** [comment]
https://www.mediawiki.org/wiki/Download_from_Git#Download_a_stable_branch has following, which is basically what we use:
```
git clone https://gerrit.wikimedia.org/r/mediawiki/core.git --branch REL1_35 mediawiki
```

**Comment by reedy** [comment]
> https://www.mediawiki.org/wiki/Download_from_Git#Download_a_stable_branch has following, which is basically what we use:
> 
> ```
> git clone https://gerrit.wikimedia.org/r/mediawiki/core.git --branch REL1_35 mediawiki
> ```

And then see the section below that :)

https://www.mediawiki.org/wiki/Download_from_Git#MediaWiki_tags_(stable_version)

>Alternatively, specific stable versions of MediaWiki are tracked using 'tags'. These are analogous to the tarball releases. Currently, these are 1.35.1 (stable), 1.31.12 (LTS) and 1.31.12 (legacy).

**Comment by tomhughes** [comment]
I've pushed https://github.com/openstreetmap/chef/commit/7bddd4d63bb8afb7a611cef67b7e9c9f3973cdf0 now to lock us to released versions so we're now on 1.35.1 which doesn't have the problem.

Of course it will come back as a problem if 1.35.2 is released with this restriction.

**Comment by matkoniecz** [comment]
> Of course it will come back as a problem if 1.35.2 is released with this restriction.

https://phabricator.wikimedia.org/T270228#6865721 has an answer:

> Probably, in a month or so.
